crates/sui-core/src/traffic_controller/policies.rs

```diff
@@ -116,13 +140,75 @@ impl TrafficSketch {
         self.sketches[self.current_sketch_index].increment(key);
     }
 
-    fn get_request_rate(&self, key: &SketchKey) -> f64 {
+    fn get_request_rate(&mut self, key: &SketchKey) -> f64 {
         let count: u32 = self
             .sketches
             .iter()
             .map(|sketch| sketch.estimate(key))
             .sum();
-        count as f64 / self.window_size.as_secs() as f64
+        let rate = count as f64 / self.window_size.as_secs() as f64;
+        self.update_highest_rates(key, rate);
+        rate
+    }
+
+    fn update_highest_rates(&mut self, key: &SketchKey, rate: f64) {
+        match key.1 {
+            ClientType::Direct => {
+                Self::update_highest_rate(
+                    &mut self.highest_rates.direct,
+                    key.0,
+                    rate,
+                    self.highest_rates.capacity,
+                );
+            }
+            ClientType::ThroughFullnode => {
+                Self::update_highest_rate(
+                    &mut self.highest_rates.proxied,
+                    key.0,
+                    rate,
+                    self.highest_rates.capacity,
+                );
+            }
+        }
+    }
+
+    fn update_highest_rate(
+        rate_heap: &mut BinaryHeap<Reverse<(IpAddr, u64)>>,
+        ip_addr: IpAddr,
+        rate: f64,
+        capacity: usize,
+    ) {
+        // Remove previous instance of this IPAddr so that we
+        // can update with new rate
+        rate_heap.retain(|&Reverse((key, _))| key != ip_addr);
+
+        let rate = rate as u64;
+        if rate_heap.len() < capacity {
+            rate_heap.push(Reverse((ip_addr, rate)));
+        } else if let Some(&Reverse((_, smallest_score))) = rate_heap.peek() {
+            if rate > smallest_score {
+                rate_heap.pop();
+                rate_heap.push(Reverse((ip_addr, rate)));
```

Comment:
if capacity is going to be largish, it will probably be better to use peek_mut, (which won't have to do upheap/downheap operations if the new element is the new smallest element). If capacity is small it may not be worth the trouble.

Reply:
Also, if you are using Reverse as the key, please construct a new Reverse and compare it, rather than comparing the rate first. The current code is confusing because you have to mentally reverse comparisons while reading the code

Reply:
Yeah it's hard coded to 20. I don't think it needs to be very large at all to ensure high accuracy of the metric